kv/kvserver/batcheval: TestDBAddSSTable failed

[(kv/kvserver/batcheval).TestDBAddSSTable failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2739208&tab=buildLog) on [master@6de4313ec216161c79fe725fcc31fc87ef1804ea](https://github.com/cockroachdb/cockroach/commits/6de4313ec216161c79fe725fcc31fc87ef1804ea):

```
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_application_decoder.go:59 +0x10f
  github.com/cockroachdb/cockroach/pkg/kv/kvserver/apply.(*Task).Decode()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/apply/task.go:123 +0x671
  github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleRaftReadyRaftMuLocked()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_raft.go:619 +0x61f
  github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleRaftReady()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_raft.go:459 +0x177
  github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).processReady()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store_raft.go:523 +0x1c8
  github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftScheduler).worker()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/scheduler.go:284 +0x3e5
  github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftScheduler).worker-fm()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/scheduler.go:229 +0x55
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:351 +0x149

Goroutine 2309 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1168 +0x5bb
  github.com/cockroachdb/cockroach/pkg/kv/kvserver/batcheval_test.TestDBAddSSTable()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/batcheval/cmd_add_sstable_test.go:76 +0x16b
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1123 +0x202

Goroutine 1613 (running) created at:
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:346 +0x107
  github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftScheduler).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/scheduler.go:205 +0x24c
  github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).processRaft()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store_raft.go:609 +0xf6
  github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store.go:1525 +0xd31
  github.com/cockroachdb/cockroach/pkg/server.(*Node).start()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:382 +0xbc4
  github.com/cockroachdb/cockroach/pkg/server.(*Server).PreStart()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:1567 +0x338f
  github.com/cockroachdb/cockroach/pkg/server.(*Server).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:1096 +0x50
  github.com/cockroachdb/cockroach/pkg/server.(*TestServer).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/testserver.go:459 +0x92
  github.com/cockroachdb/cockroach/pkg/testutils/serverutils.StartServer()
      /go/src/github.com/cockroachdb/cockroach/pkg/testutils/serverutils/test_server_shim.go:255 +0xbd
  github.com/cockroachdb/cockroach/pkg/kv/kvserver/batcheval_test.TestDBAddSSTable.func2()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/batcheval/cmd_add_sstable_test.go:83 +0x23a
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1123 +0x202
==================
    testing.go:1038: race detected during execution of test
    --- FAIL: TestDBAddSSTable/store=on-disk (1.55s)

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-race -parallel=4

```
make stressrace TESTS=TestDBAddSSTable PKG=./pkg/kv/kvserver/batcheval TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestDBAddSSTable.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/kv/kvserver/batcheval/cmd_add_sstable_test.go
+++ b/pkg/kv/kvserver/batcheval/cmd_add_sstable_test.go
@@ -100,7 +100,7 @@ func TestDBAddSSTable(t *testing.T) {
 func runTestDBAddSSTable(
 	ctx context.Context, t *testing.T, db *kv.DB, tr *tracing.Tracer, store *kvserver.Store,
 ) {
-	tr.TestingIncludeAsyncSpansInRecordings() // we assert on async span traces in this test
+	tr.TestingRecordAsyncSpans() // we assert on async span traces in this test
 	{
 		key := storage.MVCCKey{Key: []byte("bb"), Timestamp: hlc.Timestamp{WallTime: 2}}
 		data, err := singleKVSSTable(key, roachpb.MakeValueFromString("1").RawBytes)

--- a/pkg/kv/kvserver/replica_sideload_test.go
+++ b/pkg/kv/kvserver/replica_sideload_test.go
@@ -584,7 +584,7 @@ func testRaftSSTableSideloadingProposal(t *testing.T, eng storage.Engine) {
 	tc.Start(t, stopper)
 
 	tr := tc.store.ClusterSettings().Tracer
-	tr.TestingIncludeAsyncSpansInRecordings() // we assert on async span traces in this test
+	tr.TestingRecordAsyncSpans() // we assert on async span traces in this test
 	ctx, collect, cancel := tracing.ContextWithRecordingSpan(context.Background(), tr, "test-recording")
 	defer cancel()
 

--- a/pkg/kv/kvserver/replica_test.go
+++ b/pkg/kv/kvserver/replica_test.go
@@ -8487,7 +8487,7 @@ func TestFailureToProcessCommandClearsLocalResult(t *testing.T) {
 	r.mu.Unlock()
 
 	tr := tc.store.ClusterSettings().Tracer
-	tr.TestingIncludeAsyncSpansInRecordings() // we assert on async span traces in this test
+	tr.TestingRecordAsyncSpans() // we assert on async span traces in this test
 	opCtx, collect, cancel := tracing.ContextWithRecordingSpan(ctx, tr, "test-recording")
 	defer cancel()
 

--- a/pkg/util/tracing/tracer.go
+++ b/pkg/util/tracing/tracer.go
@@ -160,7 +160,8 @@ type Tracer struct {
 	// to a function that returns `true`.
 	TracingVerbosityIndependentSemanticsIsActive func() bool
 
-	includeAsyncSpansInRecordings bool // see TestingIncludeAsyncSpansInRecordings
+	testingMu               syncutil.Mutex // protects testingRecordAsyncSpans
+	testingRecordAsyncSpans bool           // see TestingRecordAsyncSpans
 
 	testing *testingKnob
 }
@@ -713,11 +714,24 @@ func (t *Tracer) VisitSpans(visitor func(*Span) error) error {
 	return nil
 }
 
-// TestingIncludeAsyncSpansInRecordings is a test-only helper that configures
+// TestingRecordAsyncSpans is a test-only helper that configures
 // the tracer to include recordings from forked/async child spans, when
 // retrieving the recording for a parent span.
-func (t *Tracer) TestingIncludeAsyncSpansInRecordings() {
-	t.includeAsyncSpansInRecordings = true
+func (t *Tracer) TestingRecordAsyncSpans() {
+	t.testingMu.Lock()
+	defer t.testingMu.Unlock()
+
+	t.testingRecordAsyncSpans = true
+}
+
+// ShouldRecordAsyncSpans returns whether or not we should include recordings
+// from async child spans in the parent span. See TestingRecordAsyncSpans, this
+// mode is only used in tests.
+func (t *Tracer) ShouldRecordAsyncSpans() bool {
+	t.testingMu.Lock()
+	defer t.testingMu.Unlock()
+
+	return t.testingRecordAsyncSpans
 }
 
 // ForkSpan forks the current span, if any[1]. Forked spans "follow from" the
@@ -734,14 +748,14 @@ func (t *Tracer) TestingIncludeAsyncSpansInRecordings() {
 //
 // [1]: Looking towards the provided context to see if one exists.
 // [2]: Unless configured differently by tests, see
-//      TestingIncludeAsyncSpansInRecordings.
+//      TestingRecordAsyncSpans.
 func ForkSpan(ctx context.Context, opName string) (context.Context, *Span) {
 	sp := SpanFromContext(ctx)
 	if sp == nil {
 		return ctx, nil
 	}
 	collectionOpt := WithParentAndManualCollection(sp.Meta())
-	if sp.Tracer().includeAsyncSpansInRecordings {
+	if sp.Tracer().ShouldRecordAsyncSpans() {
 		// Using auto collection here ensures that recordings from async spans
 		// also show up at the parent.
 		collectionOpt = WithParentAndAutoCollection(sp)

